File: plugins/echo/android/EchoPlugin/src/main/java/ca/nbsolutions/fuse/plugins/EchoPlugin.java
package ca.nbsolutions.fuse.plugins;

import android.content.res.AssetFileDescriptor;
import android.content.res.TypedArray;

import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.atomic.AtomicInteger;

import ca.nbsolutions.fuse.FuseAPIPacket;
import ca.nbsolutions.fuse.FuseAPIResponse;
import ca.nbsolutions.fuse.FuseAPIResponseStatus;
import ca.nbsolutions.fuse.FuseContext;
import ca.nbsolutions.fuse.FusePlugin;

public class EchoPlugin extends FusePlugin {
    public EchoPlugin(FuseContext context) {
        super(context);
    }

    @Override
    protected void _initHandles() {
        this.attachHandler("/echo", new APIHandler<EchoPlugin>(this) {
            @Override
            public void execute(FuseAPIPacket packet, FuseAPIResponse response) throws IOException {
                response.send(packet.readAsBinary(), packet.getContentType());
            }
        });

        this.attachHandler("/big", new APIHandler<EchoPlugin>(this) {
            @Override
            public void execute(FuseAPIPacket packet, FuseAPIResponse response) throws IOException {
                byte[] buffer = new byte[256 * 1024]; // Buffer size for reading

                response.setContentType("text/plain");
                response.setStatus(FuseAPIResponseStatus.OK);

                AssetFileDescriptor lffd = getContext().getActivityContext().getAssets().openFd("largeFile.txt");
                response.setContentLength(lffd.getLength());
                lffd.close();
                response.didFinishHeaders();

                InputStream io = getContext().getActivityContext().getAssets().open("largeFile.txt");
                int bytesRead;
                while ((bytesRead = io.read(buffer)) != -1) {
                    response.pushData(buffer);
                }

                response.didFinish();
                io.close();
            }
        });

        this.attachHandler("/subscribe", new APIHandler<FusePlugin>(this) {
            @Override
            public void execute(FuseAPIPacket packet, FuseAPIResponse response) throws IOException {
                long contentLength = packet.getContentLength();
                InputStream io = packet.getInputStream();

                byte[] buffer = new byte[(int)contentLength];
                io.read(buffer);

                String callbackID = new String(buffer, StandardCharsets.UTF_8);

                response.didFinishHeaders();
                response.didFinish();

                APIHandler<FusePlugin> self = this;

                AtomicInteger num = new AtomicInteger(0);

                Timer timer = new Timer();
                TimerTask task = new TimerTask() {
                    @Override
                    public void run() {
                        num.incrementAndGet();
                        self.plugin.getContext().execCallback(callbackID, num.toString());
                    }
                };

                timer.scheduleAtFixedRate(task, 0, 1000);
            }
        });
    }

    @Override
    public String getID() {
        return "echo";
    }
}

File: echo/ios/echo/EchoPlugin.m

/*
Copyright 2023 Breautek

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#import "EchoPlugin.h"
#import <BTFuse/BTFuse.h>

@implementation EchoPlugin

- (NSString*) getID {
    return @"echo";
}

- (void) initHandles {
    __weak EchoPlugin* weakSelf = self;
    
    [self attachHandler:@"/echo" callback:^void(BTFuseAPIPacket* packet, BTFuseAPIResponse* response) {
        NSData* message = [packet readAsBinary];
        [weakSelf doEcho: message withResponse:response];
    }];
    
    [self attachHandler:@"/echoWithReader" callback:^void(BTFuseAPIPacket* packet, BTFuseAPIResponse* response) {
        BTFuseStreamReader* reader = [[BTFuseStreamReader alloc] init: [[packet getClient] getInputStream]];
        
        NSMutableData* data = [[NSMutableData alloc] init];
        
        const int BUFFER_SIZE = 8;
        
        uint8_t buffer[BUFFER_SIZE];
        uint64_t bytesRead = 0;
        bool didError = false;
        while ((bytesRead = [reader read: buffer maxBytes: BUFFER_SIZE])) {
            if (bytesRead == -1) {
                didError = true;
                break;
            }
            
            [data appendBytes: buffer length: bytesRead];
        }
        
        [response sendData: data];
    }];
    
    [self attachHandler:@"/big" callback:^(BTFuseAPIPacket* packet, BTFuseAPIResponse* response) {
        NSString* bundlePath = [[NSBundle mainBundle] resourcePath];
        NSString* assetPath = [bundlePath stringByAppendingPathComponent:@"/assets/largeFile.txt"];
        
        NSFileManager* fileManager = [NSFileManager defaultManager];
        NSDictionary* fileAttributes = [fileManager attributesOfItemAtPath:assetPath error:nil];

        NSNumber* fileSizeNumber = [fileAttributes objectForKey:NSFileSize];
        
        [response setStatus: BTFuseAPIResponseStatusOk];
        [response setContentType:@"text/plain"];
        [response setContentLength: [fileSizeNumber unsignedIntegerValue]];
        [response didFinishHeaders];
        
        NSInputStream *inputStream = [NSInputStream inputStreamWithFileAtPath:assetPath];
        [inputStream open];

        NSUInteger bufferSize =  256 * 1024; // Adjust this according to your needs
        uint8_t buffer[bufferSize];
        NSInteger bytesRead = 0;

        while ([inputStream hasBytesAvailable]) {
            bytesRead = [inputStream read:buffer maxLength:bufferSize];
            if (bytesRead > 0) {
                NSData* data = [NSData dataWithBytes:buffer length:bytesRead];

                [response pushData:data];
            }
        }

        [inputStream close];
        
        [response didFinish];
    }];
    
    [self attachHandler:@"/subscribe" callback:^(BTFuseAPIPacket* packet, BTFuseAPIResponse* response) {
        NSString* callbackID = [packet readAsString];
        
        __block int num = 0;
        dispatch_async(dispatch_get_main_queue(), ^{
            [NSTimer scheduledTimerWithTimeInterval:1.0 repeats:true block:^(NSTimer * _Nonnull timer) {
                num++;
                
                [[self getContext] execCallback:callbackID withData:[[NSString alloc] initWithFormat:@"%d", num]];
            }];
        });
        
        [response didFinishHeaders];
        [response didFinish];
    }];
    
    [self attachHandler:@"/threadtest" callback:^(BTFuseAPIPacket* packet, BTFuseAPIResponse* response) {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
            [self doEcho: [packet readAsBinary] withResponse:response];
        });
    }];
}

- (void) doEcho:(NSData*)data withResponse:(BTFuseAPIResponse*) response {
    [response setStatus: BTFuseAPIResponseStatusOk];
    [response setContentType:@"text/plain"];
    [response setContentLength: [data length]];
    [response didFinishHeaders];
    [response pushData:data];
    [response didFinish];
}

@end
